packages/react/src/meter/root/useMeterRoot.ts

```diff
@@ -0,0 +1,179 @@
+'use client';
+import * as React from 'react';
+import { mergeReactProps } from '../../utils/mergeReactProps';
+import { valueToPercent } from '../../utils/valueToPercent';
+
+function useMeterRoot(parameters: useMeterRoot.Parameters): useMeterRoot.ReturnValue {
+  const {
+    'aria-label': ariaLabel,
+    'aria-labelledby': ariaLabelledby,
+    'aria-valuetext': ariaValuetext,
+    getAriaLabel,
+    getAriaValueText,
+    max = 100,
+    min = 0,
+    high: highParam = NaN,
+    low: lowParam = NaN,
+    optimum: optimumParam = NaN,
+    value,
+  } = parameters;
+
+  const percentageValue = valueToPercent(value, min, max);
+
+  const high = highParam ?? max;
+  const low = lowParam ?? min;
+  const optimum = optimumParam ?? (max + min) / 2;
+
+  let segment: useMeterRoot.Segment | undefined;
+
+  if (value <= low) {
+    segment = 'low';
+  } else if (value >= high) {
+    segment = 'high';
+  } else {
+    segment = 'medium';
+  }
+
+  // 'low' is preferred if `min <= optimum <= low`
+  // 'high' is preferred if `high <= optimum <= max`
+  let isOptimal = false;
+
+  if (min <= optimum && optimum <= low) {
+    isOptimal = segment === 'low';
+  } else if (high <= optimum && optimum <= max) {
+    isOptimal = segment === 'high';
+  }
+
+  const getRootProps: useMeterRoot.ReturnValue['getRootProps'] = React.useCallback(
+    (externalProps = {}) =>
+      mergeReactProps<'div'>(externalProps, {
+        'aria-label': getAriaLabel ? getAriaLabel(value) : ariaLabel,
+        'aria-labelledby': ariaLabelledby,
+        'aria-valuemax': max,
+        'aria-valuemin': min,
+        'aria-valuenow': percentageValue / 100,
+        'aria-valuetext': getAriaValueText
+          ? getAriaValueText(value)
+          : (ariaValuetext ?? `${percentageValue}%`),
+        role: 'meter',
+      }),
+    [
+      ariaLabel,
+      ariaLabelledby,
+      ariaValuetext,
+      getAriaLabel,
+      getAriaValueText,
+      max,
+      min,
+      value,
+      percentageValue,
+    ],
+  );
+
+  return {
+    getRootProps,
+    max,
+    min,
+    value,
+    percentageValue,
+    segment,
+    isOptimal,
+  };
+}
+
+namespace useMeterRoot {
+  export type Segment = 'low' | 'medium' | 'high';
```

Comment:
Have we considered going beyond the native meter capabilities and allow custom ranges? I can think of a gauge that has a preferable range in the middle. A bit lower and a bit higher would be less preferable (yellow) and the extremes on both sides could be red. The built-in range doesn't support such a scenario (but on the other hand is simpler to use in more common cases)

Reply:
Well, the whole concept of ranges is about styling, isn't it? I might not want a gradient, but have the whole indicator change color depending on which range the value is in, for example:

```css
.indicator {
  &[data-range-invalid] {
    color: red;
  }

  &[data-range-so-so] {
    color: yellow;
  }

 &[data-range-ok] {
   color: green
 }
```

where `invalid`, `so-so`, and `ok` are names defined by the developer.